### PR TITLE
check unexpected high bits for invalid characters

### PR DESCRIPTION
**Pull request by dcousens**
> 10% performance drop, but 200% better debuggability.

Hopefully this resolves that problem @indutny

### CODE REVIEW

**Comment by dcousens** [comment]
@indutny updated,  passes tests now

File: test/constructor-test.js [hunk 1]
```diff
@@ -108,6 +108,22 @@ describe('BN.js/Constructor', function () {
...
     });
+
+    it('should not accept non-hex characters', function () {
+      [
+        '0000000z',
+        '000000gg',
+        '0000gg00',
+        'ffffggff',
+        'ffffggff',
+        'hexadecimal'
```
Comment (indutny) [inline test/constructor-test.js]: Let's add `.`, `-`, `/`. They all come before `0` in ASCII.

Comment (dcousens) [reply test/constructor-test.js]: @indutny added, and it passes!

**Comment by indutny** [comment]
Sorry for delay. I just benchmarked it, and don't see any significant performance improvement. Do you see it on your side?

---

Benchmarks can be run with:
```sh
SELF_ONLY=1 node benchmarks/index.js create
```

**Comment by dcousens** [comment]
I measured an average 5-10% improvment... but the GC totally throws it off.

**Comment by fanatid** [comment]
node version: v9.2.0

master:
```
Seed: 82910582182c79d7d4d9592e15d53f0d
Benchmarking: create-10
bn.js#create-10 x 360,920 ops/sec ±2.15% (9 runs sampled)
------------------------
Fastest is bn.js#create-10
========================
Benchmarking: create-hex
bn.js#create-hex x 433,552 ops/sec ±5.14% (8 runs sampled)
------------------------
Fastest is bn.js#create-hex
========================
```

PR:
```
Seed: 5866d26a6ca6d6ce9eed6ba5f50ce0e6
Benchmarking: create-10
bn.js#create-10 x 359,505 ops/sec ±1.40% (9 runs sampled)
------------------------
Fastest is bn.js#create-10
========================
Benchmarking: create-hex
bn.js#create-hex x 459,686 ops/sec ±2.40% (9 runs sampled)
------------------------
Fastest is bn.js#create-hex
========================
```

**Comment by dcousens** [comment]
@fanatid how did you get consistent results :P

**Comment by fanatid** [comment]
Sorry, was run with different `seed`, numbers updated.

**Comment by dcousens** [comment]
@indutny good to merge?

**Pull request merged**